rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -116,6 +116,55 @@ impl CosmosProvider {
 
         Ok((account_id, signer_info.sequence))
     }
+
+    /// Calculates the sender and the nonce for the transaction.
+    /// We use `payer` of the fees as the sender of the transaction, and we search for `payer`
+    /// signature information to find the nonce.
+    /// If `payer` is not specified, we use the account which signed the transaction first, as
+    /// the sender.
+    fn sender_and_nonce(&self, tx: &Tx) -> Result<(H256, SequenceNumber), ChainCommunicationError> {
+        let (sender, nonce) = tx
+            .auth_info
+            .fee
+            .payer
+            .as_ref()
+            .map(|payer| self.search_payer_in_signer_infos(&tx.auth_info.signer_infos, payer))
+            .map_or_else(
+                || {
+                    let signer_info = tx.auth_info.signer_infos.get(0).ok_or_else(|| {
+                        ChainCommunicationError::from_other_str("no signer info in default signer")
+                    })?;
+                    self.convert_signer_info_into_account_id_and_nonce(signer_info)
+                },
+                |p| p,
+            )
+            .map(|(a, n)| CosmosAddress::from_account_id(a).map(|a| (a.digest(), n)))??;
+        Ok((sender, nonce))
+    }
+
+    /// Extract contract address from transaction.
+    /// Assumes that there is only one `MsgExecuteContract` message in the transaction
+    fn contract(tx: &Tx) -> Result<H256, ChainCommunicationError> {
+        use cosmrs::proto::cosmwasm::wasm::v1::MsgExecuteContract as ProtoMsgExecuteContract;
+
+        let any = tx
+            .body
+            .messages
+            .iter()
+            .find(|a| a.type_url == "/cosmwasm.wasm.v1.MsgExecuteContract")
+            .ok_or_else(|| {
+                ChainCommunicationError::from_other_str("could not find contract execution message")
+            })?;
+        let proto = ProtoMsgExecuteContract::from_any(any).map_err(|e| {
+            ChainCommunicationError::from_other_str(&format!(
+                "could not decode contract execution message into proto: {}",
+                e,
+            ))
+        })?;
```

Comment:
`Prost::DecodeError` is an item on `CosmosError`, so this can just be
```suggestion
        let proto = ProtoMsgExecuteContract::from_any(any)..map_err(Into::<HyperlaneCosmosError>::into)?;
```

Reply:
Fixed.